Task05/src/citiesMap.cpp

```diff
@@ -0,0 +1,134 @@
+#include "citiesMap.h"
+#include <iostream>
+#include <string>
+#include <map>
+#include <algorithm>
+#include <fstream>
+#include <sstream>
+#include <vector>
+ 
+
+void citiesMap::insert(std::string cityName, double x, double y)
+{	 
+	coordinate newCity = { x, y };
+
+	this->cityMap.insert({ cityName,newCity });
+}
+
+ 
+double calculateEuclideanDistance(double x0, double y0, double x1, double y1) {
+    return std::sqrt(std::pow(x0 - x1, 2) + std::pow(y0 - y1, 2));
+}
+
+double calculateChebyshevDistance(double x0, double y0, double x1, double y1) {
+    return std::max(std::abs(x0 - x1), std::abs(y0 - y1));
+}
+
+double calculateManhattanDistance(double x0, double y0, double x1, double y1) {
+    return std::abs(x0 - x1) + std::abs(y0 - y1);
+}
+
+void citiesMap::initilizedistanceFunctions()
+{
+     this->distanceFunctions = {
+        {0, calculateEuclideanDistance},
+        {1, calculateChebyshevDistance},
+        {2, calculateManhattanDistance}
+    };
+}
+
+
+
+void citiesMap::readCitiesFromFile(std::string filename)
+{
+   
+    std::ifstream file(filename);
+    std::string line;
+    std::string name;
+    double x, y;
+
+    while (std::getline(file, line)) {
+        name = line;
+
+        if (!std::getline(file, line)) {
+            throw std::runtime_error("Invalid data format in file: " + filename);
+        }
+
+        std::istringstream iss(line);
+        std::string xStr, yStr;
+        if (!(std::getline(iss, xStr, '-') && std::getline(iss, yStr))) {
+            throw std::runtime_error("Invalid data format in file: " + filename);
+        }
+
+        std::istringstream xIss(xStr);
+        std::istringstream yIss(yStr);
+        if (!(xIss >> x) || !(yIss >> y)) {
+            throw std::runtime_error("Invalid data format in file: " + filename);
+        }
+
+        this->insert(name, x, y);
+    }
+ 
+}
+
+bool inSquare(double x0, double y0, double x1, double y1,double radius) {
+    return   std::max(std::abs(x0 - x1), std::abs(y0 - y1)) <= radius;
+}
+
+void citiesMap::closeCities( std::string  cityName, double radius, int distanceCalc) const
+{
+    auto chosenCity  = this->cityMap.find(cityName);
+
+    if (chosenCity == this->cityMap.end()) {
+        throw std::runtime_error("City not found in the map");
+    }
+
+    if (this->distanceFunctions.find(distanceCalc) == this->distanceFunctions.end()) {
+        throw std::runtime_error("invalide distance calc function");
+    }
+   
+
+    int northCity = 0;
+
+    std::vector<std::pair<std::string,  coordinate>> closeCity;
+
+    std::copy_if(this->cityMap.begin(), this->cityMap.end(), std::back_inserter(closeCity),
```

Comment:
you are missing 1 step in finding the cities in radius, you should find vertical and horizontal rectangles, and then to find the intersection square, and to find a circle inside the square.